cel/env.go

```diff
@@ -209,13 +209,20 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	if e.chkErr != nil {
 		return nil, e.chkErr
 	}
+	decsCopy := make([]*exprpb.Decl, len(e.declarations))
+	macsCopy := make([]parser.Macro, len(e.macros))
+	progOptsCopy := make([]ProgramOption, len(e.progOpts))
+	copy(decsCopy, e.declarations)
+	copy(macsCopy, e.macros)
+	copy(progOptsCopy, e.progOpts)
+
 	ext := &Env{
-		declarations:                   e.declarations,
+		declarations:                   decsCopy,
+		macros:                         macsCopy,
+		progOpts:                       progOptsCopy,
 		adapter:                        e.adapter,
```

Comment:
Check that the adapter isn't mutable.

Reply:
It's an interface, so its mutability depends on the implementation. Our `ref.TypeRegistry` is inherently mutable, so I've added `Copy` methods to the `ref.TypeRegistry` as well as to the `pb.Db` to make it easier to ensure isolated memory. Note, the `pb.Db` copy is a shallow copy since the `FileDescription` and related objects are intended to be singletons.